lib/Mojo/IOLoop/Stream/Server.pm

```diff
@@ -0,0 +1,125 @@
+package Mojo::IOLoop::Stream::Server;
+use Mojo::Base 'Mojo::IOLoop::Stream';
+
+use Mojo::Server;
+use Mojo::Util 'term_escape';
+
+use constant DEBUG => $ENV{MOJO_SERVER_DEBUG} || 0;
+
+has app => sub { Mojo::Server->new->build_app('Mojo::HelloWorld') };
+
+sub new {
+  my $self = shift->SUPER::new(@_);
+  $self->on(read  => sub { shift->_read_content(shift) });
+  $self->on(close => sub { shift->_close });
+  return $self;
+}
+
+sub _build_tx {
+  my $self = shift;
+
+  my $tx = $self->app->build_tx;
+  $self->emit(build_tx => $tx);
+  return $tx;
+}
+
+sub _close {
+  delete($_[0]->{tx})->closed if $_[0]->{tx};
+}
+
+sub _finish {
+  my $self = shift;
+  return unless my $tx = $self->{tx};
+
+  # Finish transaction
+  delete($self->{tx})->closed;
+}
+
+sub _read_content {
+  my ($self, $chunk) = @_;
+  my $tx = $self->{tx} ||= $self->_build_tx;
+  warn term_escape "-- Server <<< Client (@{[_url($tx)]})\n$chunk\n" if DEBUG;
+  $tx->server_read($chunk);
+}
+
+sub _url { shift->req->url->to_abs }
+
+sub _write_content {
+  my $self = shift;
+
+  # Protect from resume event recursion
+  return if !(my $tx = $self->{tx}) || $self->{cont_writing};
+  local $self->{cont_writing} = 1;
+  my $chunk = $tx->server_write;
+  warn term_escape "-- Server >>> Client (@{[_url($tx)]})\n$chunk\n" if DEBUG;
+  my $next
+    = $tx->is_finished ? '_finish' : length $chunk ? '_write_content' : undef;
+  return $self->write($chunk) unless $next;
+  $self->write($chunk => sub { shift->$next() });
+}
+
+1;
+
+=encoding utf8
+
+=head1 NAME
+
+Mojo::IOLoop::Stream::Server - Non-blocking I/O server base class
+
+=head1 SYNOPSIS
+
+  package Mojo::IOLoop::Stream:MyProtoServer;
+  use Mojo::Base 'Mojo::IOLoop::Stream::Server';
+  
+  sub process { ... };
+
+=head1 DESCRIPTION
+
+Base class for building transaction based protocol specific servers.
+
+=head1 EVENTS
+
+L<Mojo::IOLoop::Stream::Server> inherits all events from
+L<Mojo::IOLoop::Stream> and can emit the following new ones.
+
+=head2 build_tx
+
+  $stream->on(build_tx => sub {
+    my ($stream, $tx) = @_;
+    ...
+  });
+
+Emitted when a transaction for new request was build.
+
+=head1 ATTRIBUTES
+
+L<Mojo::IOLoop::Stream::Server> inherits all attributes from
+L<Mojo::IOLoop::Stream> and implements the following ones.
+
+=head2 app
+
+  my $app = $stream->app;
+  $stream = $stream->app(Mojolicious->new);
+
+Holds an object responsible for building transactions. It should provide
+C<build_tx> method. Defaults to L<Mojo::HelloWorld> app.
```

Comment:
Bad description, this should always be a `Mojolicious` app, no more, no less. Just like `Mojo::Server::app`.